// el-select-tree/lib/cjs/components/ElSelectTreeVirtual.js
/*!
 * el-select-tree v2.1.1-beta.15
 * (c) 2019-2024 yujinpan
 * Released under the MIT License.
 */

'use strict';

var vuePropertyDecorator = require('vue-property-decorator');
var ElSelectTree = require('./ElSelectTree.js');
var getElSelectVirtual = require('./ElSelectVirtual.js');
var getElTreeVirtual = require('./ElTreeVirtual.js');
var utils_js = require('./utils.js');
var virtualList_js = require('./virtual-list.js');
var __vue_normalize__ = require('vue-runtime-helpers/dist/normalize-component.js');

function _applyDecoratedDescriptor(i, e, r, n, l) {
  var a = {};
  return Object.keys(n).forEach(function (i) {
    a[i] = n[i];
  }), a.enumerable = !!a.enumerable, a.configurable = !!a.configurable, ("value" in a || a.initializer) && (a.writable = !0), a = r.slice().reverse().reduce(function (r, n) {
    return n(i, e, r) || r;
  }, a), l && void 0 !== a.initializer && (a.value = a.initializer ? a.initializer.call(l) : void 0, a.initializer = void 0), void 0 === a.initializer && (Object.defineProperty(i, e, a), a = null), a;
}
function _arrayLikeToArray(r, a) {
  (null == a || a > r.length) && (a = r.length);
  for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e];
  return n;
}
function _arrayWithoutHoles(r) {
  if (Array.isArray(r)) return _arrayLikeToArray(r);
}
function _assertThisInitialized(e) {
  if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return e;
}
function _callSuper(t, o, e) {
  return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e));
}
function _classCallCheck(a, n) {
  if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(e, r) {
  for (var t = 0; t < r.length; t++) {
    var o = r[t];
    o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o);
  }
}
function _createClass(e, r, t) {
  return r && _defineProperties(e.prototype, r), Object.defineProperty(e, "prototype", {
    writable: !1
  }), e;
}
function _defineProperty(e, r, t) {
  return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
    value: t,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : e[r] = t, e;
}
function _getPrototypeOf(t) {
  return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) {
    return t.__proto__ || Object.getPrototypeOf(t);
  }, _getPrototypeOf(t);
}
function _inherits(t, e) {
  if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(e && e.prototype, {
    constructor: {
      value: t,
      writable: !0,
      configurable: !0
    }
  }), Object.defineProperty(t, "prototype", {
    writable: !1
  }), e && _setPrototypeOf(t, e);
}
function _initializerDefineProperty(e, i, r, l) {
  r && Object.defineProperty(e, i, {
    enumerable: r.enumerable,
    configurable: r.configurable,
    writable: r.writable,
    value: r.initializer ? r.initializer.call(l) : void 0
  });
}
function _isNativeReflectConstruct() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {}));
  } catch (t) {}
  return (_isNativeReflectConstruct = function () {
    return !!t;
  })();
}
function _iterableToArray(r) {
  if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r);
}
function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function ownKeys(e, r) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(e);
    r && (o = o.filter(function (r) {
      return Object.getOwnPropertyDescriptor(e, r).enumerable;
    })), t.push.apply(t, o);
  }
  return t;
}
function _objectSpread2(e) {
  for (var r = 1; r < arguments.length; r++) {
    var t = null != arguments[r] ? arguments[r] : {};
    r % 2 ? ownKeys(Object(t), !0).forEach(function (r) {
      _defineProperty(e, r, t[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {
      Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
    });
  }
  return e;
}
function _possibleConstructorReturn(t, e) {
  if (e && ("object" == typeof e || "function" == typeof e)) return e;
  if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined");
  return _assertThisInitialized(t);
}
function _setPrototypeOf(t, e) {
  return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) {
    return t.__proto__ = e, t;
  }, _setPrototypeOf(t, e);
}
function _toConsumableArray(r) {
  return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread();
}
function _toPrimitive(t, r) {
  if ("object" != typeof t || !t) return t;
  var e = t[Symbol.toPrimitive];
  if (void 0 !== e) {
    var i = e.call(t, r );
    if ("object" != typeof i) return i;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (String )(t);
}
function _toPropertyKey(t) {
  var i = _toPrimitive(t, "string");
  return "symbol" == typeof i ? i : i + "";
}
function _unsupportedIterableToArray(r, a) {
  if (r) {
    if ("string" == typeof r) return _arrayLikeToArray(r, a);
    var t = {}.toString.call(r).slice(8, -1);
    return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0;
  }
}

var _dec, _dec2, _dec3, _dec4, _class, _class2, _descriptor;
var ElSelectTreeVirtual = (_dec = vuePropertyDecorator.Component({
  name: 'ElSelectTreeVirtual',
  directives: {
    virtualList: virtualList_js.virtualList
  }
}), _dec2 = vuePropertyDecorator.Prop({
  type: Number,
  default: 200
}), _dec3 = vuePropertyDecorator.Watch('data', {
  immediate: true
}), _dec4 = vuePropertyDecorator.Watch('_expandedKeys', {
  immediate: true,
  deep: true
}), _dec(_class = (_class2 = /*#__PURE__*/function (_ElSelectTree) {
  function ElSelectTreeVirtual() {
    var _this;
    _classCallCheck(this, ElSelectTreeVirtual);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, ElSelectTreeVirtual, [].concat(args));
    _initializerDefineProperty(_this, "filterDebounceDuration", _descriptor, _this);
    _defineProperty(_this, "virtualStore", new virtualList_js.VirtualStore({
      expandedKeys: [],
      sourceData: [],
      itemHeight: 26,
      valueProp: 'value',
      childrenProp: 'children'
    }));
    _defineProperty(_this, "virtualExpandedKeys", []);
    _defineProperty(_this, "filterState", {
      stop: false
    });
    return _this;
  }
  _inherits(ElSelectTreeVirtual, _ElSelectTree);
  return _createClass(ElSelectTreeVirtual, [{
    key: "onDataChange",
    value: function onDataChange() {
      this.virtualStore.setOptions({
        sourceData: this.data,
        valueProp: this.propsMixin.value,
        childrenProp: this.propsMixin.children
      });
    }
  }, {
    key: "onExpandedKeysChange",
    value: function onExpandedKeysChange(val) {
      this.virtualStore.setOptions({
        expandedKeys: val
      });
    }
  }, {
    key: "render",
    value: function render(h) {
      this.renderValidate();
      return h(getElSelectVirtual(), this._getSelectVNodeData(), [].concat(_toConsumableArray(this.renderSlots(h)), [this.renderCacheOptions(h), h(getElTreeVirtual(), this._getTreeVNodeData())]));
    }
  }, {
    key: "_expandedKeys",
    get: function get() {
      return this.expandedKeys.concat(this.virtualExpandedKeys);
    }
  }, {
    key: "_getSelectVNodeData",
    value: function _getSelectVNodeData() {
      var _this2 = this;
      var data = this.getSelectVNodeData();
      data.directives = [{
        name: 'virtualList',
        value: {
          target: '.el-scrollbar__wrap',
          virtualStore: this.virtualStore
        }
      }];
      data.on = _objectSpread2(_objectSpread2({}, data.on), {}, {
        'visible-change': function visibleChange(val) {
          _this2._visibleChange(val);

          // reset when no selected
          if (val) {
            _this2.virtualStore.setScrollTop(0);
          }
        }
      });
      return data;
    }
  }, {
    key: "_getTreeVNodeData",
    value: function _getTreeVNodeData() {
      var _this3 = this;
      var data = this.getTreeVNodeData();
      data.props = _objectSpread2(_objectSpread2({}, data.props), {}, {
        defaultExpandedKeys: this._expandedKeys,
        defaultCheckedKeys: this.values.concat(this.defaultCheckedKeys || [])
      });
      data.props.data = this.virtualStore.data;
      data.props.dataVirtual = this.data;
      Object.assign(data.on, {
        'node-expand': function nodeExpand(node) {
          for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
            args[_key2 - 1] = arguments[_key2];
          }
          var value = _this3.getValByProp('value', node);
          if (!_this3.virtualExpandedKeys.includes(value)) {
            _this3.virtualExpandedKeys.push(value);
            utils_js.toArr(_this3.$listeners['node-expand']).forEach(function (item) {
              return item.apply(void 0, [node].concat(args));
            });
          }
        },
        'node-collapse': function nodeCollapse(node) {
          for (var _len3 = arguments.length, args = new Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
            args[_key3 - 1] = arguments[_key3];
          }
          var value = _this3.getValByProp('value', node);
          if (_this3.virtualExpandedKeys.includes(value)) {
            utils_js.spliceItem(_this3.virtualExpandedKeys, value);
            var children = _this3.getValByProp('children', node);
            if (utils_js.isValidArr(children)) {
              utils_js.treeEach(children, function (node) {
                var value = _this3.getValByProp('value', node);
                if (_this3.virtualExpandedKeys.includes(value)) {
                  utils_js.spliceItem(_this3.virtualExpandedKeys, value);
                }
              }, function (node) {
                return _this3.getValByProp('children', node);
              });
            }
            utils_js.toArr(_this3.$listeners['node-collapse']).forEach(function (item) {
              return item.apply(void 0, [node].concat(args));
            });
          }
        }
      });
      return data;
    }
  }, {
    key: "_filterMethod",
    value: function _filterMethod() {
      var _this4 = this;
      var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
      if (this.filterMethod) return this.filterMethod(val);
      this.filterState.stop = true;
      if (this.debounceTimeId) {
        clearTimeout(this.debounceTimeId);
      }
      if (!val) {
        this.virtualExpandedKeys = [];
        return this.virtualStore.setOptions({
          sourceData: this.data
        });
      }
      this.debounceTimeId = setTimeout(function () {
        utils_js.treeFilter(_this4.data, function (node) {
          return _this4._filterNodeMethod(val, node, node);
        }, _this4.propsMixin.children, _this4.filterState = {
          stop: false
        }).then(function (data) {
          var firstLeaf = utils_js.treeFind(data, function (node) {
            return !utils_js.isValidArr(node.children);
          }, function (node) {
            return _this4.getValByProp('children', node);
          });
          var firstLeafValue = firstLeaf ? _this4.getValByProp('value', firstLeaf) : undefined;
          _this4.virtualExpandedKeys = firstLeafValue ? [firstLeafValue].concat(_toConsumableArray(utils_js.getParentKeys([firstLeafValue], data, _this4.getValByProp))) : [];
          _this4.virtualStore.setOptions({
            sourceData: data
          });
        }, function () {
          // ignore
        });
      }, this.filterDebounceDuration);
    }
  }]);
}(ElSelectTree), (_descriptor = _applyDecoratedDescriptor(_class2.prototype, "filterDebounceDuration", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
}), _applyDecoratedDescriptor(_class2.prototype, "onDataChange", [_dec3], Object.getOwnPropertyDescriptor(_class2.prototype, "onDataChange"), _class2.prototype), _applyDecoratedDescriptor(_class2.prototype, "onExpandedKeysChange", [_dec4], Object.getOwnPropertyDescriptor(_class2.prototype, "onExpandedKeysChange"), _class2.prototype)), _class2)) || _class);

/* script */
var __vue_script__ = ElSelectTreeVirtual;

/* template */

/* style */
var __vue_inject_styles__ = undefined;
/* scoped */
var __vue_scope_id__ = undefined;
/* module identifier */
var __vue_module_identifier__ = undefined;
/* functional template */
var __vue_is_functional_template__ = undefined;
/* style inject */

/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = /*#__PURE__*/__vue_normalize__({}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, undefined, undefined, undefined);

module.exports = __vue_component__;
