// el-select-tree/lib/es/components/ElSelect.js
/*!
 * el-select-tree v2.1.1-beta.15
 * (c) 2019-2024 yujinpan
 * Released under the MIT License.
 */

import Vue from 'vue';

function getElSelect() {
  if (getElSelect._cache) return getElSelect._cache;
  var ElSelect = Vue.component('ElSelect');
  return getElSelect._cache = {
    extends: ElSelect,
    components: {
      ElSelectMenu: getElSelectMenu()
    }
  };
}
getElSelect._cache = null;
function getElSelectMenu() {
  if (getElSelectMenu._cache) return getElSelectMenu._cache;
  var ElSelect = Vue.component('ElSelect');
  var ElSelectMenu = ElSelect.options.components.ElSelectMenu;
  return getElSelectMenu._cache = {
    extends: ElSelectMenu,
    render: function render(h) {
      return h('div', {
        class: [{
          'el-select-dropdown': true,
          'el-popper': true,
          'is-multiple': this.$parent.multiple
        }, this.popperClass],
        style: {
          minWidth: this.minWidth
        }
      }, [this.$parent.$scopedSlots.header && h('div', {
        staticClass: 'el-select-dropdown__header'
      }, this.$parent.$scopedSlots.header()), this.$scopedSlots.default(), this.$parent.$scopedSlots.footer && h('div', {
        staticClass: 'el-select-dropdown__footer'
      }, this.$parent.$scopedSlots.footer())]);
    }
  };
}
getElSelectMenu._cache = null;

export { getElSelect as default, getElSelectMenu };
