// el-select-tree/lib/es/components/virtual-list.js
/*!
 * el-select-tree v2.1.1-beta.15
 * (c) 2019-2024 yujinpan
 * Released under the MIT License.
 */

import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime/helpers/esm/createClass';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import 'core-js/modules/es.array.concat.js';
import 'core-js/modules/es.array.includes.js';
import 'core-js/modules/es.array.push.js';
import 'core-js/modules/es.object.to-string.js';
import 'core-js/modules/es.string.includes.js';
import 'core-js/modules/web.dom-collections.for-each.js';
import throttle from 'lodash/throttle';
import { isValidArr, getCompoundVal } from './utils.js';

var virtualList = {
  inserted: function inserted(el, bindings, vNode) {
    var _bindings$value = bindings.value,
      target = _bindings$value.target,
      virtualStore = _bindings$value.virtualStore;
    var targetElem = el.querySelector(target);
    virtualStore.mount(targetElem);
    targetElem.prepend(virtualStore.sketchTopElem);
    targetElem.append(virtualStore.sketchBottomElem);
    var handleScroll = function handleScroll() {
      var old = targetElem.scrollTop;
      virtualStore.updateScroll(targetElem.scrollTop, targetElem.clientHeight, function () {
        var _vNode$componentInsta;
        targetElem.scrollTop = old;
        (_vNode$componentInsta = vNode.componentInstance) === null || _vNode$componentInsta === void 0 ? void 0 : _vNode$componentInsta.$nextTick(function () {
          targetElem.scrollTop = old;
        });
      });
    };
    targetElem.addEventListener('scroll', handleScroll);
  }
};
var VirtualStore = /*#__PURE__*/function () {
  function VirtualStore(options) {
    _classCallCheck(this, VirtualStore);
    _defineProperty(this, "data", []);
    _defineProperty(this, "sketchTopElem", document.createElement('div'));
    _defineProperty(this, "sketchBottomElem", document.createElement('div'));
    _defineProperty(this, "scrollTop", 0);
    _defineProperty(this, "clientHeight", 0);
    this.options = options;
    this.updateScroll = throttle(this.updateScroll, 15);
  }
  return _createClass(VirtualStore, [{
    key: "setOptions",
    value: function setOptions(options) {
      this.setScrollTop(options.sourceData && this.options.sourceData !== options.sourceData ? 0 : undefined);
      Object.assign(this.options, options);
      this.updateScroll();
    }
  }, {
    key: "setScrollTop",
    value: function setScrollTop() {
      var scrollTop = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.scrollTop;
      this.scrollTop = scrollTop;
      if (this.scrollElem) this.scrollElem.scrollTop = scrollTop;
      this.updateScroll();
    }
  }, {
    key: "mount",
    value: function mount(el) {
      this.scrollElem = el;
    }
  }, {
    key: "updateScroll",
    value: function updateScroll() {
      var _this = this;
      var scrollTop = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.scrollTop;
      var clientHeight = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.clientHeight;
      var callback = arguments.length > 2 ? arguments[2] : undefined;
      this.scrollTop = scrollTop;
      this.clientHeight = clientHeight || this.options.itemHeight * 15;
      var result = [];
      var height = 0;
      var heightTop = 0;
      var heightBottom = 0;
      var add = function add(node) {
        var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : result;
        var minHeight = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : scrollTop - _this.options.itemHeight * 3;
        var maxHeight = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : scrollTop + _this.clientHeight + _this.options.itemHeight * 3;
        height += _this.options.itemHeight;
        var nodeChildren = node[_this.options.childrenProp] || [];
        var newNode = _objectSpread({}, node);
        var newChildren = newNode[_this.options.childrenProp] = [].concat(nodeChildren);
        if (height < minHeight) {
          heightTop += _this.options.itemHeight;
        } else if (height >= minHeight && height - _this.options.itemHeight < maxHeight) {
          data.push(newNode);
        } else {
          heightBottom += _this.options.itemHeight;
        }
        if (isValidArr(newChildren)) {
          var _this$options$expande;
          if ((_this$options$expande = _this.options.expandedKeys) !== null && _this$options$expande !== void 0 && _this$options$expande.includes(getCompoundVal(newNode, _this.options.valueProp))) {
            newChildren.length = 0;
            nodeChildren.forEach(function (child) {
              return add(child, newChildren,
              // fix parent will show when scroll fast
              scrollTop - _this.options.itemHeight * 15);
            });
            if (newChildren.length && !data.includes(newNode)) {
              heightTop -= _this.options.itemHeight;
              height += _this.options.itemHeight;
              data.push(newNode);
            }
          } else {
            // show less children when expanded
            newChildren.length = Math.min(Math.ceil(_this.clientHeight / _this.options.itemHeight), newChildren.length);
          }
        }
      };
      this.options.sourceData.forEach(function (item) {
        return add(item);
      });
      this.data = result;
      this.sketchTopElem.style.height = heightTop + 'px';
      this.sketchBottomElem.style.height = heightBottom + 'px';
      callback === null || callback === void 0 ? void 0 : callback();
    }
  }]);
}();

export { VirtualStore, virtualList };
